### PR TITLE
feat(fetch): add programmatic proxy

### PR DESCRIPTION
**Pull request by chibat**
Releated Issue: #6654 (It has been closed..)

I think we need a way to set a proxy individually.

```ts
const client = Deno.createHttpClient({ proxy: { url: "http://myproxy.com:8080" } });
const response = await fetch("https://myserver.com", { client });
```

### CODE REVIEW

**Comment by kitsonk** [comment]
We really try to align to the WHATWG standard for Fetch and the init is supposed to just be passed through to `Request`: https://fetch.spec.whatwg.org/#fetch-method and the `RequestInit` does not include such a property, which would mean this would be specific to Deno and not something that we would consider without significant discussion _before_ considering a PR.

That being said, the standard includes a concept of a [client](https://fetch.spec.whatwg.org/#concept-request-client) but which resides on the `Request` object, not something processed by the `fetch()` method directly.  There is no way of a user setting this with init options though.

In other situations we have consider Deno specific "extensions" to the standard to reside in a `deno` property on an init object, _if_ the standard ignores excess properties and after discussion to make sure we are doing the right thing.  It appears the `Request` constructor [ignores](https://fetch.spec.whatwg.org/#request-class) excess properties in the init, so having `{ deno: { client } }` might be acceptable.

@lucacasonato thoughts?

**Comment by lucacasonato** [comment]
@kitsonk This is implemented via `Deno.createHttpClient`. `Deno.HttpClient` can already be passed to `fetch` / `Request` in the `RequestInit` in a `client` property in unstable.

File: extensions/fetch/lib.rs [hunk 1]
```diff
@@ -524,6 +546,15 @@ pub fn create_http_client(
     builder = builder.add_root_certificate(cert);
   }
 
+  if let Some(proxy) = proxy {
+    let mut reqwest_proxy = reqwest::Proxy::all(&proxy.url)?;
```
Comment (lucacasonato) [inline extensions/fetch/lib.rs:550]: The proxy URL should go through a network permission check here.

Comment (lucacasonato) [reply extensions/fetch/lib.rs:550]: Or more accurately in the op_create_http_client

Comment (chibat) [reply extensions/fetch/lib.rs:556]: @kitsonk @lucacasonato Thank you for your comments.

I fixed to check a network permission.

**Review by lucacasonato — APPROVED** [review]
Looks great, no further comments. Thanks :-)

We will wait with merging until the 1.12 merge window opens later next week.

**Comment by mokeyish** [comment]
When will this feature be released?

**Comment by bartlomieju** [comment]
In 1.12 on July 13th

**Pull request merged**